apps/zotonic_core/src/models/m_rsc_import.erl

```diff
@@ -18,96 +17,1130 @@
+                        {Acc, ImpAcc}
+                end;
+            (<<"content_group">>, _, {Acc, ImpAcc}) -> {Acc, ImpAcc};
+
+            % Other ids are mapped to placeholders or local ids
+            (<<"menu">>, Menu, {Acc, ImpAcc}) when is_list(Menu) ->
+                % map ids in menu to local ids
+                {Menu1, ImpAcc1} = map_menu(Menu, UriTemplate, [], ImpAcc, Options, Context),
+                {Acc#{ <<"menu">> => Menu1 }, ImpAcc1};
+            (<<"blocks">>, Blocks, {Acc, ImpAcc}) when is_list(Blocks) ->
+                % map ids in blocks and content to local ids
+                {Blocks1, ImpAcc1} = map_blocks(Blocks, UriTemplate, [], ImpAcc, Options, Context),
+                {Acc#{ <<"blocks">> => Blocks1 }, ImpAcc1};
+            (K, V, {Acc, ImpAcc}) ->
+                case m_rsc_export:is_id_prop(K) of
+                    true ->
+                        case map_id(V, UriTemplate, ImpAcc, OptionsNoEdges, Context) of
+                            {ok, {LocalId, ImpAcc1}} ->
+                                ImpAcc2 = ImpAcc1#{
+                                    uri(V) => LocalId
+                                },
+                                {Acc#{ K => LocalId }, ImpAcc2};
+                            {error, _} ->
+                                {Acc#{ K => undefined }, ImpAcc}
+                        end;
+                    false ->
+                        {V1, ImpAcc1} = map_html(K, V, UriTemplate, ImpAcc, OptionsNoEdges, Context),
+                        {Acc#{ K => V1 }, ImpAcc1}
+                end
+        end,
+        {#{}, ImportedAcc},
+        Rsc),
+
+    % Set forced and default props
+    Rsc2 = maps:merge(Rsc1, PropsForced),
+    Rsc3 = maps:merge(PropsDefault, Rsc2),
+    Rsc4 = case maps:find(<<"is_authoritative">>, PropsForced) of
+        {ok, true} ->
+            Rsc3#{
+                <<"is_authoritative">> => true,
+                <<"uri">> => undefined
+            };
+        _ ->
+            Rsc3#{
+                <<"uri">> => maps:get(<<"uri">>, RemoteRId),
+                <<"is_authoritative">> => false
+            }
+    end,
+
+    % Ensure that the is_published flag is present, defaults to true
+    Rsc5 = case maps:find(<<"is_published">>, Rsc4) of
+        {ok, true} -> Rsc4;
+        _ -> Rsc4#{ <<"is_published">> => true }
+    end,
+    {Rsc5, ImportedAcc1}.
+
+
+%% @doc Map all ids in a menu to local (stub) resources. Skip all tree entries
+%% that could not be mapped to local ids.
+map_menu([ #rsc_tree{ id = RemoteId, tree = Sub } | Rest ], UriTemplate, Acc, ImpAcc, Options, Context) ->
+    case map_id(RemoteId, UriTemplate, ImpAcc, Options, Context) of
+        {ok, {LocalId, ImpAcc1}} ->
+            {Sub1, ImpAcc2} = map_menu(Sub, UriTemplate, [], ImpAcc1, Options, Context),
+            Acc1 = [ #rsc_tree{ id = LocalId, tree = Sub1 } | Acc ],
+            map_menu(Rest, UriTemplate, Acc1, ImpAcc2, Options, Context);
```

Comment:
I was wondering why you would need to reverse the list at the [] case of map_menu; it is to avoid having to concatenate lists here, so simple prepends can be used. This is more efficient because single reverses on list ends cost less than appending several times within the lists.

Reply:
Because we go recursively through the lists and prepend each item the lists gets reversed in the accumulator.
That is why at the end the accumulator is reversed.